`vrt-redact-long`: Redact multiple attributes, redact structural attributes, allow counting bytes

janiemi opened the pull request:
This pull request contains a couple of additions or improvements to `vrt-redact-long` and library modules, and a modification of `vrt-validate` to recognize too long VRT lines and attribute values. The main changes are the following:

- `vrt-redact-long`:
  - Allow redacting multiple positional attributes at the same time, specified with `--field`, similarly to `vrt-keep`.
  - Add `--struct`: Specify structural attributes to be redacted.
  - Add `--bytes`: Count lengths in bytes instead of characters, preserving UTF-8 validity.
  - To improve speed, read and write in `bytes` and only convert to strings as needed.
- `vrt-validate`:
  - Error if a line is longer than 65,534 bytes (`cwb-encode` fails in error).
  - Warn if a structural or positional attribute value is longer than 4,095 bytes (`cwb-encode` truncates).
- `libvrt.metaline`:
  - Add function `escape` to encode the characters `& < > "` (in `bytes`).
  - Add functions `strunescape` and `strescape` working on strings.
  - Add functions `starttag` and `strstarttag` to construct a structure start tag form structure name and attributes.
- Add `libvrt.structargs`:
  - Parse command-line option values specifying structural attribute names similarly to `libvrt.nameargs` for positional attribute names: functions `structattrbagtype`, `structattranybagtype` and `parsestructattrs`.

Tests are unfortunately missing, except for the changes to `vrt-validate`.

janiemi commented:
@jpiitula, I now requested your review of my changes. Unfortunately, tests are missing, except for the additions to `vrt-validate` .

jpiitula commented:
I think strescape would be better written with str.translate, as str.maketrans allows strings as replacements for characters, unlike bytes.maketrans that can only replace with single bytes. Not that I've ever measured it to see if any speed difference is noticable.

jpiitula commented:
starttag and strstarttag assume that attribute values are already escaped. Is that right? Yes, maybe that is right. It might be good to mention this in the doc string.

janiemi commented:
@jpiitula, thanks for your comments this far.

`starttag` and `strstarttag` indeed don’t escape attribute values themselves, so I’ll add that to the doc string.

I got a bit mixed speed results for `str.translate` vs. `str.replace`:

- 5 characters, all to be replaced: `translate` about twice as fast as `replace`:
    ```sh
    $ python3 -m timeit -s 's = "&<>&\""' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    1000000 loops, best of 3: 0.786 usec per loop
    $ python3 -m timeit -s 's = "&<>&\""' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    1000000 loops, best of 3: 0.386 usec per loop
    ```
- 5 characters, none to be replaced: `translate` slightly faster:
    ```sh
    $ python3 -m timeit -s 's = "abcde"' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    1000000 loops, best of 3: 0.493 usec per loop
    $ python3 -m timeit -s 's = "abcde"' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    1000000 loops, best of 3: 0.418 usec per loop
    ```
- 10 characters, 5 to be replaced: effectively the same speed:
    ```sh
    $ python3 -m timeit -s 's = "<a>b&c&d\"e"' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    1000000 loops, best of 3: 0.774 usec per loop
    $ python3 -m timeit -s 's = "<a>b&c&d\"e"' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    1000000 loops, best of 3: 0.773 usec per loop
    ```
- 5000 characters, all to be replaced: `replace` about 2.5 times as fast as `translate`
    ```sh
    $ python3 -m timeit -s 's = "&<>&\"" * 1000' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    10000 loops, best of 3: 127 usec per loop
    $ python3 -m timeit -s 's = "&<>&\"" * 1000' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    1000 loops, best of 3: 323 usec per loop
    ```
- 5000 characters, none to be replaced: `translate` about 1.5 times as fast as `replace`
    ```sh
    $ python3 -m timeit -s 's = "abcde" * 1000' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    100000 loops, best of 3: 8.1 usec per loop
    $ python3 -m timeit -s 's = "abcde" * 1000' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    100000 loops, best of 3: 5.16 usec per loop
    ```
- 5000 characters, 2500 to be replaced: `replace` 5 times as fast as `translate`
    ```sh
    $ python3 -m timeit -s 's = "<a>b&c&d\"e" * 500' 's.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")'
    10000 loops, best of 3: 75.3 usec per loop
    $ python3 -m timeit -s 's = "<a>b&c&d\"e" * 500' -s 'tr = str.maketrans(dict([("&", "&amp;"),("<", "&lt;"),(">", "&gt;"),("\"", "&quot;")]))' 's.translate(tr)'
    1000 loops, best of 3: 375 usec per loop
    ```

I don’t know how much difference it would make in practice.

janiemi commented:
As you maybe noticed, I rebased the branch to split to a separate commit the change to read and write in binary instead of text. That was my original intention, but I noticed only after creating this pull request that I hadn’t done that.

jpiitula commented:
I note that parsestructattrs is quite complex. It is hard to be sure that it is not missing some corner case. It might be possible to make it simpler by making the type check more precise, maybe more strict about the placement of colons. Not sure, though.

I think for redactions the usual case would be to work on just one attribute, and only after an issue has been detected, which already should be rare. But who knows what other use cases there might be.

I don't like that it does not detect duplicates. It might be catastrophic to process an attribute twice.

But I also think that we can live with this.

jpiitula commented:
About those timings: you are timing maketrans repeatedly when it should only be executed once when the module is loaded? and surely the intended use of the mechanism is such that most of the time no replacements happen, yet this is always run on every word or such.

jpiitula commented:
Re binprefix, binsuffix, I think catching the decoding error is fine, unless a much more local mechanism can be found, and I failed to find one. (Julia uses UTF-8 directly and has methods to find the next or previous valid index at any point. I don't see that anywhere in Python.) I would not worry about efficiency, as this should not happen often anyway.

The whole redaction mechanism is rather scary anyway. Splitting between a base character and a combining character would be awkward enough, and what if we happen to remove only one of a pair of bidi markers? Live with these for now, I guess.

jpiitula commented:
I do not remember what islikeword is for. Some silly heuristic, I hope. Because otherwise the restriction of binislikeword to ASCII sounds like something that is not right.

janiemi commented:
@jpiitula, thanks again for your comments. Here are some comments on and answers to them, all in this same comment.

> I note that parsestructattrs is quite complex. It is hard to be sure that it is not missing some corner case. It might be possible to make it simpler by making the type check more precise, maybe more strict about the placement of colons.

I guess you’re right. A leading colon causes an error, whereas a trailing colon can make it effectively ignore the attribute (`text:title:`). I’ll make the regular expression in `structattrbagtype` more precise. However, I don’t immediately see how to simplify `parsestructattrs` significantly.

(I should have added tests. I still could, but I’m not sure if I should spend time on that right now. I think you were right in suggesting that I should fix the STT data by hand (perhaps a one-off Perl one-liner or something) and not modify `vrt-redact-long`, but I couldn’t resist the temptation and then got a bit carried away.)

> I think for redactions the usual case would be to work on just one attribute, and only after an issue has been detected, which already should be rare.

You’re right. For positional attributes it would probably typically be only the word form, as I think it would be better to redact before parsing and adding other attributes. However, in the case of STT, the data had already been parsed, so also the lemma had to be redacted. Moreover, for structural attributes I can imagine cases in which it would make sense to redact multiple attributes on the same run. However, the script is relatively fast, so it probably wouldn’t be a problem to have to run it multiple times.

My idea was to try to treat positional and structural attributes analogously in this respect, but I probably should have consulted you first. However, I can revert the change if you’d prefer that, at least for positional attributes.

> I don't like that it does not detect duplicates. It might be catastrophic to process an attribute twice.

Do you mean in general or in `vrt-redact-long`? As far as I can see, the redaction operation would seem idempotent. Anyway, for other uses of `parsestructattrs`, I can add duplicate detection.

> About those timings: you are timing maketrans repeatedly when it should only be executed once when the module is loaded? and surely the intended use of the mechanism is such that most of the time no replacements happen, yet this is always run on every word or such.

No, the commands have two `-s` options for setting up: one for the string to be processed and another for `maketrans`, so `maketrans` is executed only once.

> Re binprefix, binsuffix, I think catching the decoding error is fine, unless a much more local mechanism can be found, and I failed to find one.

Good. This was what I managed to come up with. I didn’t try to find out if the Python standard library would have had something for that, but I take your word that it doesn’t.

> The whole redaction mechanism is rather scary anyway. Splitting between a base character and a combining character would be awkward enough, and what if we happen to remove only one of a pair of bidi markers?

Right. I hadn’t thought about that, but then that’s not directly related to the changes I made. I suppose that it wouldn’t be too difficult to add checks for these, at least if you’re working with characters and not bytes. However, I’d let you do it as you know much more about those issues than I do.

> Live with these for now, I guess.

I agree. I think those problems are relatively marginal.

> I do not remember what islikeword is for. Some silly heuristic, I hope.

It’s used to allow redacting only “word-like” values.

> Because otherwise the restriction of binislikeword to ASCII sounds like something that is not right.

That’s true. I’ll change `binislikeword` to decode `bytes` to `str` and call `islikeword`.

janiemi commented:
One more note: neither `vrt-redact-long` nor `vrt-validate` takes into account the effect of encoding the one-byte characters `/ < > |` as two-byte characters before encoding the data with `cwb-encode`. As a result, even if the VRT data had all attribute values clearly under 4095 bytes long, if they contain those characters, the length of their encoded representation may exceed 4095 bytes, making `cwb-encode` truncate the values.

I personally think that this would be too specific to our Korp instance to take into account in the VRT tools. Moreover, I have some hopes that we could eventually get rid of that extra encoding, as CWB has now got some features that might make it unnecessary. It would require some changes in the Korp backend, though.

janiemi commented:
@jpiitula, I now pushed commits to fix a couple of issues you commented on:

- `binislikeword` works correctly with Unicode characters; that is, `--like=word --bytes` works.
- `libvrt.structargs.parsestructattrs` raises an error on leading, trailing and multiple consecutive colons, and on duplicate attribute names.
- `libvrt.structargs.structattrbagtype` checks the list of structural attributes with a more precise regexp.

The script still allows redacting multiple attributes on the same run.

I didn’t try to simplify `parsestructattrs`; on the contrary, I added some error-checking code there, too. However, I added a couple of comments, which I hope may clarify the code.

janiemi commented:
@jpiitula, I’ll now finally merge this to `master`, even though you haven’t commented on my last changes, as I’d like to use in `vrt-add-id` some of the functions I added here to `libvrt.metaline`.
